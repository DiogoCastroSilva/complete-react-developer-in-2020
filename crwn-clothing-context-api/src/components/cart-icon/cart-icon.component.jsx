import React, { useContext } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { ReactComponent as ShoppingIcon } from '../../assets/shopping-bag.svg';
import './cart-icon.styles.scss';

import { selectCartItemsCount } from '../../redux/cart/cart.selectors';

import CartContext from '../../context/cart/cart.context';


const CartIcon = ({ itemCount }) => {
  const { toggleHidden } = useContext(CartContext);

  return (
    <div className='cart-icon' onClick={toggleHidden}>
      <ShoppingIcon className='shopping-icon' />
      <span className='item-count'>{itemCount}</span>
    </div>
  );
};


const mapStateToProps = createStructuredSelector({
  itemCount: selectCartItemsCount
});

export default connect(mapStateToProps)(CartIcon);
