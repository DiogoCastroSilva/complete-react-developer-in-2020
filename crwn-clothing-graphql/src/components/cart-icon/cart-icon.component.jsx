import React from 'react';

import './cart-icon.styles.scss';

import { ReactComponent as ShoppingIcon } from '../../assets/shopping-bag.svg';


const CartIcon = ({ toggleCartHidden, itemsCount }) => (
  <div className='cart-icon' onClick={toggleCartHidden}>
    <ShoppingIcon className='shopping-icon' />
    <span className='item-count'>{itemsCount}</span>
  </div>
);

export default CartIcon;
