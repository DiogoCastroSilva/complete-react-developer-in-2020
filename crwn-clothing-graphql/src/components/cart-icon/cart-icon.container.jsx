import React from 'react';
import { Mutation, Query } from 'react-apollo';
import { gql } from 'apollo-boost';

import { GET_ITEMS_COUNT } from '../../grapql/resolvers';

import CartIcon from './cart-icon.component';


const TOGGLE_CART_HIDDEN = gql`
    mutation ToggleCartHidden {
        toggleCartHidden @client
    }
`;


const CartIconContainer = () => (
    <Query query={GET_ITEMS_COUNT}>
        {
            ({ data }) => {
                const { itemsCount } = data;
                return (
                    <Mutation mutation={TOGGLE_CART_HIDDEN}>
                        {
                            toggleCartHidden =>
                                <CartIcon
                                    toggleCartHidden={toggleCartHidden}
                                    itemsCount={itemsCount}
                                />
                        }
                    </Mutation>
                );
            }
        }
    </Query>
);

export default CartIconContainer;