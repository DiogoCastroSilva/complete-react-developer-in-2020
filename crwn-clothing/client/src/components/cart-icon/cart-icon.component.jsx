import React from 'react';

import {
    CartIconContainer,
    ShoppingItemContainer,
    ItemCountContainer
} from './cart-icon.styles';


const CartIcon = ({ toggleCartHidden, itemsCount }) => (
    <CartIconContainer onClick={toggleCartHidden}>
        <ShoppingItemContainer />
        <ItemCountContainer>
            {itemsCount}
        </ItemCountContainer>
    </CartIconContainer>
);

export default CartIcon;