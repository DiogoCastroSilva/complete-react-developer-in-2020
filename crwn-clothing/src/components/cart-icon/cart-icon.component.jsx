import React from 'react';
import { connect } from 'react-redux';

import {
    CartIconContainer,
    ShoppingItemContainer,
    ItemCountContainer
} from './cart-icon.styles';

import { toggleCartHidden } from '../../redux/cart/cart.actions';
import { selectCartItemsCount } from '../../redux/cart/cart.selectors';

const CartIcon = ({ toggleCartHidden, itemsCount }) => (
    <CartIconContainer onClick={toggleCartHidden}>
        <ShoppingItemContainer />
        <ItemCountContainer>
            {itemsCount}
        </ItemCountContainer>
    </CartIconContainer>
);

const mapDispatchToProps = dispatch => {
    return {
        toggleCartHidden: () => dispatch(toggleCartHidden())
    };
};

const mapStateToProps = state => {
    return {
        itemsCount: selectCartItemsCount(state)
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(CartIcon);