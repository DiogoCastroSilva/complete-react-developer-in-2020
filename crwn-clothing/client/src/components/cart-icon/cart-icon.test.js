import React from 'react';
import { shallow } from 'enzyme';

import CartIcon from './cart-icon.component';
import { ItemCountContainer } from './cart-icon.styles';


describe('Testing CartIcon component', () => {
    let wrapper;
    let mockToggleCartHidden;

    beforeEach(() => {
        mockToggleCartHidden = jest.fn();

        const mockProps = {
            toggleCartHidden: mockToggleCartHidden,
            itemsCount: 0
        };

        wrapper = shallow(<CartIcon {...mockProps} />);
    });

    it('expects to render CartIcon component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('expects toogle to be called', () => {
        wrapper.simulate('click');
        expect(mockToggleCartHidden).toHaveBeenCalled();
    });

    it('expects item count to be rendered', () => {
        const itemCounts = parseInt(wrapper.find(ItemCountContainer).text());
        expect(itemCounts).toBe(0);
    })
});