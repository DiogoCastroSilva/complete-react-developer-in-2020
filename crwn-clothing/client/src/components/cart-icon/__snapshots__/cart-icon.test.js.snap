// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CartIcon component expects to render CartIcon component 1`] = `
<styled.div
  onClick={[MockFunction]}
>
  <Styled(Component) />
  <styled.span>
    0
  </styled.span>
</styled.div>
`;
